cmake/detect_gpu.cmake

```diff
@@ -0,0 +1,221 @@
+                       Building for all known architectures.")
+        set(${out_variable} ${ginkgo_known_gpu_archs} PARENT_SCOPE)
+    else()
+        set(${out_variable} ${CUDA_gpu_detect_output} PARENT_SCOPE)
+    endif()
+endfunction()
+
+
+################################################################################
+# A function for automatic detection of GPUs installed
+# (if autodetection is enabled)
+# Usage:
+#     ginkgo_detect_nvcc_version(out_variable)
+function(ginkgo_detect_nvcc_version out_variable)
+    set(__nvfile ${PROJECT_BINARY_DIR}/detect_nvcc_version.cu)
+    file(WRITE ${__nvfile} ""
+        "#include <cstdio>\n"
+        "int main()\n"
+        "{\n"
+        "  int count = 0;\n"
+        "  if (cudaSuccess != cudaGetDeviceCount(&count)) return -1;\n"
+        "  if (count == 0) return -1;\n"
+        "  std::printf(\"%d\", __CUDACC_VER_MAJOR__);\n"
+        "  return 0;\n"
+        "}\n")
+    execute_process(COMMAND "${CMAKE_CUDA_COMPILER}" "--run" "${__nvfile}"
+                    WORKING_DIRECTORY "${PROJECT_BINARY_DIR}/CMakeFiles/"
+                    RESULT_VARIABLE __nv_res OUTPUT_VARIABLE __nv_out
+                    ERROR_QUIET OUTPUT_STRIP_TRAILING_WHITESPACE)
+    if(__nv_res EQUAL 0)
+        set(${out_variable} ${__nv_out} PARENT_SCOPE)
+    endif()
+endfunction()
+
+
+################################################################################
+# A function for automatic detection of GPUs installed 
+# (if autodetection is enabled)
+# Usage:
+#     check_available(arch_var,nvcc_var)
+function(check_available arch_ver nvcc_ver)
+    if (${arch_ver} LESS 30)
+        message(FATAL_ERROR "Do not support gpu architecture < 30")
+    elseif ((${arch_ver} GREATER_EQUAL 70) AND (${nvcc_ver} LESS 9))
+        message(FATAL_ERROR "Volta support: Needs cuda version >= 9.0")
+    endif()
+endfunction()
+
+################################################################################
+# Function for selecting GPU arch flags for nvcc based on CUDA_ARCH_NAME
+# Usage:
+#     ginkgo_select_nvcc_arch_flags(out_variable)
+function(ginkgo_select_nvcc_arch_flags out_variable)
+    set(__nvcc_ver 8) # default cuda 8
+    ginkgo_detect_nvcc_version(__nvcc_ver)
+
+    # List of arch names
+    if (__nvcc_ver GREATER_EQUAL 9)
+        set(__archs_names "Kepler" "Maxwell" "Pascal" "Volta" "All" "Manual")
+        list(APPEND known_gpu_archs "70")
+    else()
+        set(__archs_names "Kepler" "Maxwell" "Pascal" "All" "Manual")
+    endif()
+
+    set(__archs_name_default "Auto")
+    if(NOT CMAKE_CROSSCOMPILING)
+        list(APPEND __archs_names "Auto")
+        set(__archs_name_default "Auto")
+    endif()
+
+    # set CUDA_ARCH_NAME strings (so it will be seen as dropbox in CMake-Gui)
+    set(CUDA_ARCH_NAME ${__archs_name_default} CACHE STRING
+        "Select target NVIDIA GPU achitecture.")
+    set_property( CACHE CUDA_ARCH_NAME PROPERTY STRINGS "" ${__archs_names} )
```

Comment:
So what if I want to compile for multiple architectures? For example, both `Kepler` and `Pascal`. I cannot do that here, as I can only set 1 architecture name, and not a list of them.

Also, I might want to do: `Maxwell;Auto`, to compile for `Maxwell`, and anything else that was automatically detected. Or specifically, for all Maxwells, and 6.0 (maybe by setting `Maxwell;60`).
I also don't see the need for `Manual`, why wouldn't the user just set one variable to `60` instead of having to set two variables (one to `Manual`, and the other to `60`)?

This can be elegantly solved by looping through the input list of values, and just copying architectures to the output list, unless we find a keyword like `Kepler`, which is expanded to a list of `Kepler` architectures, or `Auto` which is expanded to a list of automatically detected architectures.